File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast3.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        var ret = new List<int>();
        var stack = new Stack<TreeNode>();

        bool GoLeft() {
            for (; root != null; root = root.left) stack.Push(root);
            return stack.Count != 0;
        }

        for (; GoLeft(); root = root.right) {
            root = stack.Pop();
            ret.Add(root.val);
        }

        return ret;
    }
}

File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ret = []
        stack = []

        while True:
            while root is not None:
                stack.append(root)
                root = root.left

            if not stack:
                break

            root = stack.pop()
            ret.append(root.val)
            root = root.right

        return ret
